### ISSUE TITLE
Candidate Address and Address Override 

### ISSUE THREAD

**Issue by Justin-Pickett** [department-of-veterans-affairs/va.gov-team#1287]
Candidate Address and Address Override

[VA Profile Onboarding Guide](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/identity-personalization/change-address/address-validation/Vet360_Address_Validation_Onboarding_Guide_v1.4%20(3).docx)

Candidate Address is going out in the next Vet360 build. The UX work has been covered in #11339.

In addition, Vet360 is shipping Address Override. This is intended as a feature of last resort: if the Pitney Bowes Spectrum address database has not yet caught up to the cartographical realities of the US housing market, the Veteran should be able to say "yes, this is actually a real address."

Address Override needs a UX which places it at the end of various checks, eg Candidate Address and

"Are you absolutely sure???"

This feature is currently available from Vet360.

The original EPIC is in vets.gov-team repo [#15034] (https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/vets.gov-team/15034)

[GitHub Documentation]( )

We'll use this epic to track the progress for combining the existing Profile and Account pages.

## Project management

- [x] Create product outline [#2256](https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/2256)

## Design
- [x] Design Discovery 
- [x] Design Wireframes/Prototypes - [1688](https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/1688)
- [x] [Content review](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/content/content-review-process.md#how-to-request-content-review)
  - Copy was written and approved by Peggy, but is in a mess of tickets not documented here.
- [x] [Design review](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/design/working-with-platform-design-team.md)
  - Design was reviewed with Samara and Shawna and does not need additional review.

##  Design Research / User Testing 
- [x] Research Plan #2326
- [x] Socialize research plan with appropriate team members for approval (Lauren Alexanderson)
- [x] Create Conversation Guide and Schedule Review #2327
- [x] User Testing Sessions Scheduled #2326
- [x] Conduct user testing sessions #3280
- [x] Create research summary #3629
- [x] Create tickets for any changes/ updates that came from user testing #3741

## Build Tickets

- [x] [Front-End] Web version -Candidate Address Validation #1700
- [x] [FE] research: determine how to call the validation API instead of the update/save endpoint from the address edit modal #3172
- [x] [FE] Make address comparison function #3173
- [x] [FE] Adjust the call to update/save #3174
- [x] [FE] Create redux actions/reducers #3175
- [x] [FE] Create modal/visual components #3176
- [x] [FE] Determine possible test scenarios #3180
- [x]  Use only candidate api for address validation api #4060
- [x] Address validation: Continue, Cancel, and X buttons are broken #4205
- [x] Address validation: Copy is missing a period #4211
- [x]  Validation Keys do not work to override addresses in the profile #4245
- [x]  Do not show the user any suggestions that are not "CONFIRMED" as deliverable #4250
- [x] The address validation modal remembers the last option you selected between opens and closes #4264
- [x] User should see the address they entered when going back to edit their address #4271
- [x] Sometimes we get validation keys but can't force the addresses through #4278
- [x] Review the flow for Address Validation Modal #4283
- [x] 1st attempt to force an overridable address works, 2nd+ fail #4327
- [x] In some cases picking a suggested address from the API will not work. #4328
- [x] In some cases address updates fail because we make a POST call instead of a PUT #4246
- [x] Address selection radio button isn't visible #4373
- [x] Raise address confidence threshold from 80 to 90 #4468
- [x] Address suggestion bug #4487
- [x] Disable "update" button if no address is selected #4483
- [x] Update for address validation where there is only one option #4632
- [x] Address Validation Error State not resetting #4837
- [x] Address validation: Implement the new endpoint from VA Profile #5080
- [x] Address validation - can't select entered address with validation key #5178
- [x] Address validation - Still getting validation keys that don't work #5180

## QA/Testing

- [x] Quality Assurance testing #3948
  - [x] Create test plan. #5010
  - [x] Run QA and add bug tickets based on QA. #5012
  - [x] #5466 (Alert-message copy bug)
  - [x] #5505 (IE11 radio-button quirk)
  - [x] #5553 (Firefox alert-icon styling bug)
- [x] [E2E tests](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/quality-assurance/e2e-testing/README.md), running in CI/CD, passing on all browsers #5005
- [x] [Load testing](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/quality-assurance/load-testing/README.md) #5032

## 508/accessibility

- [x] [Accessibility / 508 review](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/teams/vsa/accessibility/vsa-accessibility-review-process.md) #5323
- [x] [SCREENREADER, KEYBOARD]: Edit address link will not receive keyboard focus #5866
- [x] [SCREENREADER, KEYBOARD]--PART 2: Edit address link will not receive keyboard focus #6008
- [x] [SCREENREADER, KEYBOARD]: Focus SHOULD return to Edit button that launched the modals when modals are closed #6156
- [x] [SCREENREADER]: Focus is not being returned to the Edit mailing address button on Cancel from confirmation modal #6470

## Analytics

- [x] Determine what analytics we need for this feature #5705
- [x] Analytics implementation #5895
- [x] Address Validation Analytics - Front End Implementation #6213

## Pre-launch

[Full pre-launch checklist](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/working-with-vsp/onboarding/Product%20Development%20Checklist.md)

- [x] Create call center guide #5084
- [x] [Create a release plan](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/product-management/release-plan-template.md) #5085
- [x] [PM] Update product outline pre-launch #5094
- [x] [BE] Address validation: Review product outline and update if needed #5254
- [x] [FE] Address validation: Review product outline and update if needed #5256
- [x] [Downtime UX and error messaging documentation complete](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/identity-personalization/change-address/address-validation/error-code-matrix.md)
- [x] [Security review](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/security/security-review-process.md)/Privacy review/ Production readiness review [41](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/issues/41)
- [x] Monitoring requirements ([see product outline](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/identity-personalization/change-address/address-validation/product-outline.md#how-to-access-and-test))
- [x] Get/implement production API key #6207
- [x] "Learn and Improve" plan written for Address Validation feature #6212
- [x] Address Validation: bad handling of 400s from address_validation endpoint #6203
- [x] Address Validation modal does not handle failure to create address save transaction #6696
- [x] Go/No Go #6725
- [x] Launch! #6426

## Post-launch

- [x] [SCREENREADER]: Set focus on the update message, then back to Edit when user updates their profile mailing address #6472
- [x] [FE] Address validation: Analytics fix #7054
- [x] Change address update endpoint to use iso3 country code #6046
- [x] [BE] Stop using country name for VA profile contact information endpoint #5591
- [x] Follow up with veterans to see if this feature resolves their issues saving addresses #6012
- [x] Full Accessibility & 508 Office Audit [Address Validation] #8006

Backlog/follow-on work

- [x] [FE/BE] What is causing the mailing and home address errors we are still seeing? #7532
- [x] Investigate why some VA.gov profiles are failing to initialize in VA Profile [101](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/issues/101)

**Comment by Justin-Pickett** [comment]
Design discovery happening.

Technical discovery needed as well.

Lihan will check into how this connects with Vets360.

Get email addy's from Samara/Lisa.

**Comment by Justin-Pickett** [comment]
** Notes of call with Vets360 team on 8/30/19 **

** VFT / VA.gov Questions for Vet360: **

In terms of user flow, when should the candidate address endpoint be used, compared to validate address endpoint?

_"Ian stated Address Validation is more about someone typing in an address to ensure it’s valid or getting one from some other source like an address mailed in by someone on behalf of a Veteran." _

* If a candidate address isn’t quite right, up to 3 suggestions may be presented to users. *

**Can you provide example request/response for address override?
Ian will forward this question to Amy Rosenthal. **

_ "Erik stated that without seeing design mocks, it’s unclear what VFT needs, but it’s good to know what option we have." _

*Lisa would like VFT to provide multiple addresses for the user to pick from, or the ability to override this and keep the originally-input address. *

* Arthur advised VFT’s current designs show that when a user edits an address, the current address is shown in the form. If a different address is input, the UI currently shows what the user entered along with multiple suggested addresses for selection. *

_The design illustrates the ability for the user to type a PO box or other unrecognizable current address and override the system’s suggestion_ 

** VFT team asked how many addresses Vet360 returns, and Ian reached out to Vet360’s Rachel Fulton for details:**
* The maximum addresses the service will return is 10 *
* If the address in the request is an exact match it will return 1 address *
* If the service cannot figure out the address request at all, it will return 0 addresses *

_ Vet360 uses a Pitney-Bowes product for addresses _

** Lihan suggested VFT should use one Vet360 endpoint that will return multiple addresses, if an exact match isn’t required **

_ Erik advised they may have to do back / front end filtering if they get exact matches. Can deal with this on their side _

_ Arthur stated he’ll get information based on Lisa’s questions, and from their work through the UX with Lihan and at that point maybe will have more questions for Vet360 _

**Comment by artistgreen** [comment]
Notes from Meeting w/ @lihanli @erikphansen and @mojaray2k :

- We don't have definitive idea of what will be returned for addresses from Vet360
- Lihan is waiting on authorization for his API Key requests to be able to determine what is returned
- We know Vet360 can return up to 10 address suggestions but are asking the question: 'how may suggestions should we show?'
- Based on Lihan's technical discovery, we will know if the returned suggested addresses will be in any order (e.g., order of confidence score)
- Erik made the suggestion to visually call out any suggested address differences from the user entered address

**Comment by artistgreen** [comment]
Screenshots with 1st revisions from initial review discussions - 
![Profile-suggest an address-profile-rev1](https://user-images.githubusercontent.com/53535009/65174114-3f74bd00-da1e-11e9-88cd-52d9531fce18.png)

![Profile-suggest an address-edit address-rev1](https://user-images.githubusercontent.com/53535009/65174171-5adfc800-da1e-11e9-844e-dcee86cc582e.png)

![Profile-suggest an address-waiting message-rev1](https://user-images.githubusercontent.com/53535009/65174195-64693000-da1e-11e9-8d1e-50a88e513625.png)

![Profile-suggest an address- address error-rev1](https://user-images.githubusercontent.com/53535009/65174216-6d5a0180-da1e-11e9-8337-9dea2e72bc63.png)

![Profile-suggest an address-address not recognized message-rev1](https://user-images.githubusercontent.com/53535009/65174233-7945c380-da1e-11e9-92ff-f41ec94ad528.png)

![Profile-suggest an address-address not recognized updated-rev1](https://user-images.githubusercontent.com/53535009/65174292-a09c9080-da1e-11e9-87b7-bd5324cd818a.png)

![Profile-suggest an address-suggested address updated 1-rev1](https://user-images.githubusercontent.com/53535009/65174306-a7c39e80-da1e-11e9-8d6f-6f737f360d4c.png)

![Profile-suggest an address-updated profile-rev1](https://user-images.githubusercontent.com/53535009/65174326-af834300-da1e-11e9-98fd-8c948bacd28c.png)

**Comment by lkoenigsberg** [comment]
Refer to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/11416 for some additional technical API chatter, if helpful.

**Comment by erikphansen** [comment]
# Address validation response handling acceptance criteria
The front end should handle different responses from the address validation/suggestion endpoint (`POST v0/profile/address_validation`) as follows:

1. If the response is a single deliverable suggestion (`"delivery_point_validation": "CONFIRMED"`) with a `"confidence_score"` greater than 80, the user should not see any suggestions. Instead the address the user has entered should be sent directly to the save/update endpoint and the address change flow will appear to function exactly like it does today.
2. If the response is a single deliverable suggestion with a confidence score of less than 80, show the user the suggestion. If the response included a `"validation_key"` then we should also give the user the ability to use the address that they had originally entered.
3. If the response is an error with a validation key, tell the user that the address appears to be incorrect/invalid and prompt them to go back and edit what they entered, but allow them to force it through.
4. If the response is an error _without_ a validation key, tell the user that the address appears to be incorrect/invalid and prompt them to go back and edit what they entered. Do not give them the option to force the address through. Instead present them with a help desk phone number to call if they are sure the address they entered was correct (even though the help desk can't do much at this point).
5. If the response is a single suggestion with a `"delivery_point_validation"` value of `"STREET_NUMBER_VALIDATED_BUT_MISSING_UNIT_NUMBER"` or `"STREET_NUMBER_VALIDATED_BUT_BAD_UNIT_NUMBER"`, tell the user that they appear to have left off the unit number or entered an invalid unit number and prompt them to edit. But also give them the option to force the address through.
6. If the response returns multiple suggestions, only show them the `"CONFIRMED"` addresses. Also prompt them to go back and edit if none of the suggestions are their address. Finally, if there is validation key, allow them to force through the address that they entered. But if there isn't a validation key, present them with a help desk phone number to call if they are sure the address they entered was correct (even though the help desk can't do much at this point).
7. In any other case (I'm not sure what cases are missed here) tell the user that the address does not appear to be correct and prompt them to go back and edit. Also present them with a help desk phone number.

## Example request/responses
### For case 1
```
{
  "address": {
    "address_line1": "225 irving st #1",
    "city": "San Francisco",
    "address_pou": "RESIDENTIAL",
    "address_type": "DOMESTIC",
    "country_name": "United States",
    "state_code": "CA",
    "zip_code": "94123"
  }
}

{
  "addresses": [
    {
      "address": {
        "address_line1": "225 Irving St Apt 1",
        "address_type": "DOMESTIC",
        "city": "San Francisco",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "06075",
        "county_name": "San Francisco",
        "state_code": "CA",
        "zip_code": "94122",
        "zip_code_suffix": "2633"
      },
      "address_meta_data": {
        "confidence_score": 88.0,
        "address_type": "Domestic",
        "delivery_point_validation": "CONFIRMED",
        "residential_delivery_indicator": "RESIDENTIAL"
      }
    }
  ],
  "validation_key": 95384771
}
```

### For case 2
```
{
  "address": {
    "address_line1": "110 washingtonn",
    "city": "portland",
    "state_code": "or",
    "zip_code": "97215",
    "zip_code_suffix": ""
  }
}

{
  "addresses": [
    {
      "address": {
        "address_line1": "110 SE Washington St",
        "address_type": "DOMESTIC",
        "city": "Portland",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "41051",
        "county_name": "Multnomah",
        "state_code": "OR",
        "zip_code": "97214",
        "zip_code_suffix": "2127"
      },
      "address_meta_data": {
        "confidence_score": 77.0,
        "address_type": "Domestic",
        "delivery_point_validation": "CONFIRMED",
        "residential_delivery_indicator": "BUSINESS"
      }
    }
  ],
  "validation_key": 1272603069
}
```

### For case 3
```
{
  "address": {
    "address_line1": "washingtonn",
    "city": "portland",
    "state_code": "or",
    "zip_code": "97215",
    "zip_code_suffix": ""
  }
}

{
  "errors": [
    {
      "title": "Address Validation Error",
      "detail": {
        "address": {
          "state_province": {
            "name": "Oregon",
            "code": "OR"
          },
          "country": {
            "name": "USA",
            "code": "USA",
            "fips_code": "US",
            "iso2_code": "US",
            "iso3_code": "USA"
          },
          "address_line1": "Washingtonn",
          "city": "Portland",
          "zip_code5": "97215"
        },
        "geocode": {
          "calc_date": "2019-11-14T15:44:09+00:00",
          "location_precision": 0.0,
          "latitude": 45.5151,
          "longitude": -122.5999
        },
        "address_meta_data": {
          "confidence_score": 0.0,
          "address_type": "Unknown",
          "delivery_point_validation": "MISSING_ZIP",
          "validation_key": 457300252
        },
        "messages": [
          {
            "code": "ADDRVAL112",
            "key": "AddressCouldNotBeFound",
            "severity": "ERROR",
            "text": "The Address could not be found"
          }
        ]
      },
      "code": "VET360_AV_ERROR",
      "status": "400"
    }
  ]
}
```

### For case 5
```
{
  "address": {
    "address_line1": "225 irving st",
    "city": "San Francisco",
    "state_code": "CA",
    "zip_code": "94122",
    "zip_code_suffix": ""
  }
}

{
  "addresses": [
    {
      "address": {
        "address_line1": "225 Irving St",
        "address_type": "DOMESTIC",
        "city": "San Francisco",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "06075",
        "county_name": "San Francisco",
        "state_code": "CA",
        "zip_code": "94122",
        "zip_code_suffix": "2652"
      },
      "address_meta_data": {
        "confidence_score": 98.0,
        "address_type": "Domestic",
        "delivery_point_validation": "STREET_NUMBER_VALIDATED_BUT_MISSING_UNIT_NUMBER",
        "residential_delivery_indicator": "RESIDENTIAL"
      }
    }
  ],
  "validation_key": 2125510284
}
```

### For case 6
```
{
  "address": {
    "address_line1": "400 65th St",
    "city": "seattle",
    "state_code": "WA",
    "zip_code": "12345"
  }
}

{
  "addresses": [
    {
      "address": {
        "address_line1": "400 N 65th St",
        "address_type": "DOMESTIC",
        "city": "Seattle",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "53033",
        "county_name": "King",
        "state_code": "WA",
        "zip_code": "98103",
        "zip_code_suffix": "5252"
      },
      "address_meta_data": {
        "confidence_score": 100.0,
        "address_type": "Domestic",
        "delivery_point_validation": "UNDELIVERABLE"
      }
    },
    {
      "address": {
        "address_line1": "400 NW 65th St",
        "address_type": "DOMESTIC",
        "city": "Seattle",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "53033",
        "county_name": "King",
        "state_code": "WA",
        "zip_code": "98117",
        "zip_code_suffix": "5026"
      },
      "address_meta_data": {
        "confidence_score": 100.0,
        "address_type": "Domestic",
        "delivery_point_validation": "CONFIRMED",
        "residential_delivery_indicator": "RESIDENTIAL"
      }
    },
    {
      "address": {
        "address_line1": "400 NE 65th St",
        "address_type": "DOMESTIC",
        "city": "Seattle",
        "country_name": "USA",
        "country_code_iso3": "USA",
        "county_code": "53033",
        "county_name": "King",
        "state_code": "WA",
        "zip_code": "98115",
        "zip_code_suffix": "6463"
      },
      "address_meta_data": {
        "confidence_score": 98.0,
        "address_type": "Domestic",
        "delivery_point_validation": "STREET_NUMBER_VALIDATED_BUT_MISSING_UNIT_NUMBER",
        "residential_delivery_indicator": "RESIDENTIAL"
      }
    }
  ],
  "validation_key": 178464079
}
```

**Comment by artistgreen** [comment]
@erikphansen - Awesome and thanks for the info!

Sounds like the front end design covers every response scenario except for no. 5:

5.   If the response is a single suggestion with a "delivery_point_validation" value of 
     "STREET_NUMBER_VALIDATED_BUT_MISSING_UNIT_NUMBER" or 
     "STREET_NUMBER_VALIDATED_BUT_BAD_UNIT_NUMBER", tell the user that they appear to 
     have left off the unit number or entered an invalid unit number and prompt them to edit. But 
     also give them the option to force the address through.

There were some comments regarding the possibility of showing the user error text on the form requesting that the user enter or re-enter their 'Unit number'.

@erikphansen - any thoughts on that?   Thanks!

**Comment by erikphansen** [comment]
That would depend on how you wanted to add the error text to the form modal. Putting an alert component on the page would likely be the easiest. Placing some sort of call out by the street field itself might be trickier. @bradl3yC might have some thoughts since he's been digging into the relevant code recently.

**Comment by Samara-Strauss** [comment]
From a UI perspective, it sounds like we have UI for the following use cases:

- Address is invalid, but override is possible. Suggestions are also available.
- Address is invalid, but override is possible. No suggestions are available.
- Address is invalid, but override is NOT possible. Suggestions are also available.
- Address is invalid, but override is NOT possible. Suggestions are not available.

It sounds like we may need UI updates/copy for the following use case:

- Address is determined to be invalid because of a bad/missing unit number. We want to suggest here that they go back and edit their address to fix the unit number issue.
  - @erikphansen in this case, I assume that it is also possible we will have addresses to suggest to them. Is this assumption correct or incorrect?

@artistgreen @erikphansen is my summary of needed states accurate? If so, we need to work on the bad/missing unit number use case.

**Comment by erikphansen** [comment]
@Samara-Strauss in the case of a missing/bad unit number, there are no suggestions to make since the suggestion API can't/doesn't suggest addresses _with_ unit numbers.

Other than that everything you said is correct.

**Comment by Samara-Strauss** [comment]
@erikphansen thank you for clarifying. All clear on that.

@artistgreen can you please modify the UI as needed and work with copy resources for the use case where there are bad/missing unit numbers?

**Comment by artistgreen** [comment]
@Samara-Strauss - yes, will work on that - thanks.

**Comment by artistgreen** [comment]
Hi @peggygannon - this is the ticket I was referring to in my email regarding content for the 'bad/missing unit number' modal.  Thanks!

**Comment by Samara-Strauss** [comment]
@artistgreen please create a new ticket for @peggygannon. It is not clear what you need from her based on this ticket, and there are too many comments here and it's too long to ask her to sift through everything to find what she needs. A new ticket for your copy needs would be kinder and clearer.

I'm closing this. Copy updates from user testing still need to happen but they are being tracked here #3741.

**Comment by peggygannon** [comment]
@artistgreen Hello. I wanted to double-check that you need me to provide copy for when a user needs to enter a military address, as we discussed? Is there a content ticket for that work? This issue here is an epic with many different issues discussed.

**Comment by artistgreen** [comment]
@peggygannon - Hi, yes, I just tagged you on the ticket for the military base copy.  Thank you!

**Comment by Samara-Strauss** [comment]
All follow-up tickets have been completed. Closing! CC/ @mshea0606 for awareness.
